### PR TITLE
Add Publish_Docs stage, make it push to a fake gh-pages branch for testing whenever master is updated.

### PR DESCRIPTION
**Pull request by nlewycky**
bors r-

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -210,6 +210,29 @@ jobs:
...
+      - template: .azure/install-cmake.yml
+      - bash: |
+          sudo apt-get install doxygen graphviz
+        displayName: Install doxygen
+      - bash: |
+          make docs
+        displayName: Build docs
+      - publish: $(System.DefaultWorkingDirectory)/api-docs
+        artifact: api-docs
+        displayName: Publish doc artifact
```
Comment (Hywan) [inline azure-pipelines.yml]: Should we rephrase “Publish doc artifact”  to “Move [or Save] doc artifact”? I would reserve the word “publish” to pushing online/on `gh-pages`.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -259,27 +282,34 @@ jobs:
...
-        submodules: true
-      - template: .azure/install-rust.yml
-      - template: .azure/install-llvm.yml
-      - template: .azure/install-sccache.yml
-      - template: .azure/install-cmake.yml
+      - task: DownloadPipelineArtifact@2
+        inputs:
+          artifactName: api-docs
+          targetPath: $(System.DefaultWorkingDirectory)/api-docs
+          # targetPath  : $(Pipeline.Workspace)/api-docs
```
Comment (Hywan) [inline azure-pipelines.yml]: Do we need to keep this comment?

File: azure-pipelines.yml (hunk 1 above)
Comment (Hywan) [reply azure-pipelines.yml]: doc Also, do we need to move the artifacts for `stating` and `trying`? I don't think so, maybe just for `master`?

Comment (syrusakbary) [reply azure-pipelines.yml]: Cool. I will name it `Save doc artifact`.

Note: I have a plan for reusing the doc artifacts for creating nicer files for the C libraries :)

**Comment by syrusakbary** [comment]
> One question: Do we want the online documentation to reflect the state of the master branch, or the releases only?

I think we can build it always on `master`. Happy to change it if you have some preference here!

File: Makefile [hunk 3]
```diff
@@ -303,7 +301,21 @@ dep-graph:
...
+	mkdir -p api-docs/c
+	cp -R target/doc api-docs/rust
+	cp -R lib/runtime-c-api/doc/html api-docs/c/runtime-c-api
+	echo '<!-- Build $(SOURCE_VERSION) --><meta http-equiv="refresh" content="0; url=rust/wasmer_runtime/index.html">' > api-docs/index.html
+	echo '<!-- Build $(SOURCE_VERSION) --><meta http-equiv="refresh" content="0; url=wasmer_runtime/index.html">' > api-docs/rust/index.html
+
+docs-publish:
+	git clone -b "gh-pages" --depth=1 https://wasmerbot:$(GITHUB_DOCS_TOKEN)@github.com/wasmerio/wasmer.git api-docs-repo
+	cp -R api-docs/* api-docs-repo/
+	cd api-docs-repo && git add index.html rust/* c/*
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply Makefile]: We decided to go with:


```
api-docs/
    crates/
        wasmer-runtime-core/ (rustdoc)
        wasmer-llvm-backend/ (rustdoc)
        …/ (rustdoc)
    c/
        wasmer-runtime-c-api/ (doxygen)
```

**Comment by Hywan** [comment]
We can say that `wasmerio.github.io/api-docs/` contains latest documentation, and docs.rs serves the “official” per-release API docs. I think it's a good strategy.

**Pull request merged**